README.md
# Snowbirdie

> A [Snowbird](https://wordpress.org/themes/snowbird/) child theme.

## Table of Contents

* [Installation](#installation)
* [Changes](#changes)
  * [Actions](#actions)
  * [Additions](#additions)
  * [Style](#style)
  * [Template Parts](#template-parts)
* [Demo](#demo)
* [License](#license)

## Installation

Snowbirdie provides a `composer.json` file so you can install it via [Composer](https://getcomposer.org/):

```shell
$ composer require tfrommen/snowbirdie
```

You can also just [use the generated ZIP file](https://github.com/tfrommen/snowbirdie/releases).

**Notice:** Since Snowbirdie is a **child theme**, you need to have [Snowbird](https://wordpress.org/themes/snowbird/) installed as well.

## Changes

The following sections list all changes (i.e., customizations as well as additions) that will be put in effect when installing Snowbirdie.

### Actions

#### `snowbirdie.after_footer_widgets`

Snowbirdie provides the `snowbirdie.after_footer_widgets` action hook that is fired right after the footer widgets. Hooking there allows you to inject markup after the footer widgets.

#### `snowbirdie.before_footer_widgets`

Snowbirdie provides the `snowbirdie.before_footer_widgets` action hook that is fired right before the footer widgets. Hooking there allows you to inject markup before the footer widgets.

### Additions

#### Author Bio Sidebar

Snowbirdie provides an additional widget area, _Author Bio_ (`'author-bio'`), that is located right after the author bio.

#### Footer Widgets

Snowbirdie provides a full-width search input in the footer, just before the widgets.

![Search in Footer](images/search-in-footer.jpg)

### Style

#### Author Bio

Snowbirdie wraps the author bio in `<p />` tags by using `wpautop()`.

Furthermore, the whole author bio section is pushed below the content when in mobile view.

#### Avatar

Snowbirdie uses rounded borders on avatars.

![Avatar](images/avatar.jpg)

#### Color Scheme

Snowbirdie provides an additional color scheme: _snowbirdie_.

![Snowbirdie Color Scheme](images/color-scheme.jpg)

#### Content Width

Snowbirdie uses a content width of 1280 pixels (instead of the original 1110 pixels).

#### Excerpt

Snowbirdie displays a Read More button below every excerpt, no matter if generated or not.

![Read More Button](images/read-more.jpg)

#### Footer Text

Snowbirdie renders a footer text that consists of a copyright notice for the current year, the title of the site, and the description.

![Footer Text](images/footer-text.jpg)

#### Image Sizes

Snowbirdie uses the following customized image sizes (all cropped center|center):

* Featured Image: 1920 x 720
* _Snowbird Large_: 1280 x 480
* _Snowbird Thumb_: 400 x 400
* _Snowbird Small_: 120 x 120

#### Links

Snowbirdie decorates all links inside posts with an underline.

### Template Parts

#### Featured Image

Snowbirdie displays a `<figcaption />` element containing the featured image's caption, if exists.

![Featured Image Caption](images/featured-image-caption.jpg)

#### Loop Sidebar

Snowbirdie always lists the author (i.e., avatar and name) of each post, no matter if there only is one author on the site.

#### Post Meta

For single post view, Snowbirdie also includes the number of comments in the post meta section.

![Comments in Post Meta](images/comments-in-post-meta.jpg)

## Demo

If you would like to see Snowbirdie in action, please refer to [my personal website](https://tfrommen.de).

## License

This code is licensed under the [GNU General Public License](LICENSE).


CHANGELOG.md
# Changelog

## 1.1.0

* Official GitHub release.

## 1.0.0

* Initial release.


template-parts/post-meta.php
<?php
/**
 * Display for only Post
 */
if ( 'post' === get_post_type() ) : ?>
	<div class="xf__meta" itemprop="text">
		<?php
		/**
		 * Icons for Sticky, Private/Protected Post
		 */
		if ( is_sticky() && is_home() && ! is_paged() ) : ?>
			<span class="xf__meta-item sticky"><i class="fa fa-thumb-tack"></i> </span>
		<?php elseif ( post_password_required() || 'private' === get_post_status() ) : ?>
			<span class="xf__meta-item private"><i class="fa fa-lock"></i> </span>
		<?php endif; ?>

		<?php
		/**
		 * Post Categories
		 */
		$categories_list = get_the_category_list( esc_html_x( ', ', 'category separator', 'snowbird' ) );

		if ( $categories_list ) : ?>
			<span class="xf__meta-item link-category"><?php echo $categories_list; ?></span>
		<?php endif; ?>

		<?php
		/**
		 * Post Time - Published on
		 */
		printf(
			'<span class="xf__meta-item posted-on"><a href="%1$s" rel="bookmark"><time class="entry-date published" datetime="%2$s" itemprop="datePublished">%3$s</time></a></span>',
			esc_url( get_the_permalink() ),
			esc_attr( get_the_time( 'c' ) ),
			esc_html( get_the_time( get_option( 'date_format' ) ) )
		);
		/* Post Time - Updated On */
		printf(
			'<span class="xf__meta-item updated-on screen-reader-text"><time class="entry-date updated" datetime="%1$s" itemprop="dateModified">%2$s</time></span>',
			esc_attr( get_the_modified_time( 'c' ) ),
			esc_html( get_the_modified_time( get_option( 'date_format' ) ) )
		);
		/**
		 * Comment Count
		 */
		if ( is_single() && ! post_password_required() ) : ?>
			<span class="xf__meta-item comment-count">
				<?php comments_popup_link(
					esc_html__( 'Add Comment', 'snowbird' ),
					esc_html__( '1 Comment', 'snowbird' ),
					esc_html__( '% Comments', 'snowbird' ),
					'',
					''
				); ?>
			</span>
		<?php endif; ?>
	</div>
<?php endif;


template-parts/loop-sidebar.php
<div class="xf__entry-sidebar clear">

	<div class="xf__entry-meta">
		<?php echo get_avatar( get_post()->post_author, '85' ); ?>
		<p class="entry-author vcard" itemtype="http://schema.org/Person" itemscope="itemscope" itemprop="author">
			<span><?php esc_html_e( 'by:', 'snowbird' ); ?></span>
			<span class="fn author" itemprop="name"><?php echo snowbird_get_author(); ?></span></p>

		<?php if ( ! post_password_required() ) : ?>
			<p class="comment-count">
				<?php comments_popup_link(
					esc_html__( 'Add Comment', 'snowbird' ),
					esc_html__( '1 Comment', 'snowbird' ),
					esc_html__( '% Comments', 'snowbird' ),
					'',
					''
				); ?>
			</p>
		<?php endif; ?>

		<hr class="separator">
	</div>

	<?php
	/**
	 * Share this post
	 */
	$thumbnail = '';
	$thumb     = wp_get_attachment_image_src( get_post_thumbnail_id(), 'large' );

	if ( isset( $thumb[0] ) ) {
		$thumbnail = $thumb[0];
	}

	/**
	 * Check for registered action, if available
	 * bail to the action callback.
	 *
	 * @see https://developer.wordpress.org/reference/functions/add_action/
	 */
	if ( has_action( 'snowbird_share_this_loop' ) ) :
		do_action( 'snowbird_share_this_loop', $thumbnail );
	else : ?>
		<div class="xf__entry-share">
			<p class="share-title"><?php esc_html_e( 'Share on:', 'snowbird' ); ?></p>

			<p>
				<a class="facebook" title="<?php esc_html_e( 'Share on Facebook', 'snowbird' ); ?>"
				   href="https://www.facebook.com/sharer.php?u=<?php echo urlencode( esc_url( get_the_permalink() ) ) ?>&amp;t=<?php echo urlencode( esc_attr( get_the_title() ) ) ?>">
					<i class="fa fa-facebook"></i>
					<span class="screen-reader-text"><?php esc_html_e( 'Facebook', 'snowbird' ); ?></span></a>
				<a class="twitter" title="<?php esc_html_e( 'Tweet This', 'snowbird' ); ?>"
				   href="https://twitter.com/share?text=<?php echo urlencode( esc_attr( get_the_title() ) ) ?>&amp;url=<?php echo urlencode( esc_url( get_the_permalink() ) ) ?>">
					<i class="fa fa-twitter"></i>
					<span class="screen-reader-text"><?php esc_html_e( 'Twitter', 'snowbird' ); ?></span></a>
				<a class="pinterest" title="<?php esc_html_e( 'Pin This', 'snowbird' ); ?>"
				   href="https://www.pinterest.com/pin/create/button/?url=<?php echo urlencode( esc_url( get_the_permalink() ) ) ?>&amp;description=<?php echo urlencode( esc_attr( get_the_title() ) ) ?>&amp;media=<?php echo urlencode( esc_url( $thumbnail ) ) ?>">
					<i class="fa fa-pinterest-p"></i>
					<span class="screen-reader-text"><?php esc_html_e( 'Pinterest', 'snowbird' ); ?></span></a>
				<a class="gplus" title="<?php esc_html_e( 'Share on Google+', 'snowbird' ); ?>"
				   href="https://plus.google.com/share?url=<?php echo urlencode( esc_url( get_the_permalink() ) ) ?>">
					<i class="fa fa-google-plus"></i>
					<span class="screen-reader-text"><?php esc_html_e( 'Google+', 'snowbird' ); ?></span></a>
				<a class="linkedin" title="<?php esc_html_e( 'Share on LinkedIn', 'snowbird' ); ?>"
				   href="https://www.linkedin.com/shareArticle?mini=true&amp;title=<?php echo urlencode( esc_attr( get_the_title() ) ) ?>&amp;url=<?php echo urlencode( esc_url( get_the_permalink() ) ) ?>">
					<i class="fa fa-linkedin"></i>
					<span class="screen-reader-text"><?php esc_html_e( 'LinkedIn', 'snowbird' ); ?></span></a>
				<a class="mail" title="<?php esc_html_e( 'Email This', 'snowbird' ); ?>"
				   href="mailto:?subject=<?php echo urlencode( esc_attr( get_the_title() ) ) ?>&amp;body=<?php echo urlencode( esc_url( get_the_permalink() ) ) ?>">
					<i class="fa fa-envelope-o"></i>
					<span class="screen-reader-text"><?php esc_html_e( 'Email', 'snowbird' ); ?></span></a>
			</p>
		</div>
		<?php
	endif;

	/**
	 * Make use of the spot for something cool.
	 *
	 * @see https://developer.wordpress.org/reference/functions/add_action/
	 */
	if ( has_action( 'snowbird_loop_sidebar' ) ) :
		do_action( 'snowbird_loop_sidebar' );
	endif; ?>
</div>


template-parts/featured-image.php
<?php
/**
 * Featured Image
 */
if ( has_post_thumbnail() && ! post_password_required() && ! is_attachment() ) : ?>
	<figure class="xf__feaured-image" aria-hidden="true">
		<?php
		if ( is_singular() ) {

			the_post_thumbnail( 'post-thumbnail', array(
				'alt'      => get_the_title(),
				'itemprop' => 'image'
			) );

			$post_tumbnail = get_post( get_post_thumbnail_id() );
			if ( ! empty( $post_tumbnail->post_excerpt ) ) {
				printf(
					'<figcaption>%s</figcaption>',
					wp_kses_post( $post_tumbnail->post_excerpt )
				);
			}

		} else {

			printf(
				'<a href="%s">%s</a>',
				esc_url( get_the_permalink() ),
				get_the_post_thumbnail( get_the_ID(), 'post-thumbnail', array(
					'alt'      => get_the_title(),
					'itemprop' => 'image'
				) )
			);

		} ?>
	</figure>
<?php endif;


functions.php
<?php # -*- coding: utf-8 -*-

/**
 * Renders the footer widgets.
 *
 * @since 1.0.0
 *
 * @return void
 */
function snowbird_display_footer_widgets() {

	$columns = snowbird_maybe_display_footer();
	if ( $columns || has_action( 'snowbirdie.before_footer_widgets' ) || has_action( 'snowbirdie.after_footer_widgets' ) ) {
		?>
		<div class="widget-area">
			<?php
			/**
			 * Fires right before the footer widgets.
			 *
			 * @since 1.0.0
			 */
			do_action( 'snowbirdie.before_footer_widgets' );

			if ( $columns ) : ?>
				<div class="xf__container row clear">
					<?php $classes = snowbird_get_footer_widget_classes(); ?>
					<?php for ( $i = 1; $i <= $columns; $i++ ) : ?>
						<?php if ( is_active_sidebar( "footer-$i" ) ) : ?>
							<div class="<?php echo esc_attr( $classes . ( 1 === $i ? ' first' : '' ) . ( $columns === $i ? ' last' : '' ) ); ?>">
								<?php dynamic_sidebar( "footer-$i" ); ?>
							</div>
						<?php endif; ?>
					<?php endfor; ?>
				</div>
			<?php endif;

			/**
			 * Fires right after the footer widgets.
			 *
			 * @since 1.0.0
			 */
			do_action( 'snowbirdie.after_footer_widgets' );
			?>
		</div>
		<?php
	}
}

/**
 * Returns the author bio.
 *
 * @since 1.0.0
 *
 * @param bool $generate Optional. Generate bio if empty? Defaults to false.
 *
 * @return string Author bio.
 */
function snowbird_get_author_bio( $generate = false ) {
	$bio = get_the_author_meta( 'description', get_post()->post_author );

	if ( $generate && empty( $bio ) ) {
		$post_count = count_user_posts( get_post()->post_author );

		// Translators: 1: author 2: total posts count
		$bio = sprintf(
			esc_html__( '%1$s has been contributed to a whooping %2$s.', 'snowbird' ),
			get_the_author_meta( 'display_name', get_post()->post_author ),
			sprintf( _n( '%d article', '%d articles', number_format_i18n( $post_count ), 'snowbird' ), number_format_i18n( $post_count ) )
		);
	}

	return $bio ? wpautop( $bio ) : '';
}

add_filter( 'snowbird_content_width', function () {

	return 1280;
} );

add_filter( 'snowbird_get_color_schemes', function ( array $color_schemes ) {

	return array_merge( [
		'snowbirdie' => [
			'label'  => esc_html_x( 'snowbirdie', 'admin', 'snowbirdie' ),
			'colors' => [
				'header_text_color'        => '#0a0909',
				'header_background_color'  => '#fcfbfa',
				'content_title_color'      => '#401b0a',
				'content_text_color'       => '#424242',
				'content_alt_text_color'   => '#696969',
				'content_accent_color'     => '#ff5505',
				'content_background_color' => '#fbfbfb',
				'footer_title_color'       => '#fff6f2',
				'footer_text_color'        => '#faf6f5',
				'footer_alt_text_color'    => '#fff6f2',
				'footer_accent_color'      => '#ff5505',
				'footer_background_color'  => '#0d0907',
				'button_text_color'        => '#fcfbfa',
				'button_background_color'  => '#401b0a',
			],
		],
	], $color_schemes );
} );

add_filter( 'get_the_excerpt', function ( $output ) {

	if ( has_excerpt() && ! is_admin() && ! is_attachment() ) {
		return $output . snowbird_filter_excerpt_more();
	}
} );

add_filter( 'snowbird_filter_excerpt_more', function () {

	return '<p><a class="xf__more xf__button" href="' . esc_url( get_permalink() ) . '">' . esc_html__( 'Read More', 'snowbird' ) . '</a></p>';
} );

add_filter( 'snowbird_footer_text', function () {

	$site_name = get_bloginfo( 'name' );

	return sprintf(
		esc_html__( 'Copyright &copy; %1$s %2$s // %3$s', 'snowbirdie' ),
		date( 'Y' ),
		sprintf(
			'<a href="%2$s" title="%3$s" rel="home">%1$s</a>',
			esc_html( $site_name ),
			esc_url( home_url( '/' ) ),
			esc_attr( $site_name )
		),
		esc_html( get_bloginfo( 'description' ) )
	);
} );

add_action( 'init', function () {

	remove_action( 'after_setup_theme', 'snowbird_logo_data_update', 99 );
	remove_action( 'after_setup_theme', 'snowbird_jetpack_setup' );
	remove_action( 'tgmpa_register', 'snowbird_recommended_plugins' );
} );

add_action( 'after_setup_theme', function () {
	
	set_post_thumbnail_size( 1920, 720, true );

	add_image_size( 'snowbird-large', 1280, 480, true );
	add_image_size( 'snowbird-thumb', 400, 400, true );
	add_image_size( 'snowbird-small', 120, 120, true );
}, 20 );

add_action( 'widgets_init', function () {
	
	register_sidebar( [
		'name'          => esc_html_x( 'Author Bio', 'admin', 'snowbirdie' ),
		'id'            => 'author-bio',
		'description'   => 'Sidebar right below the author bio, displayed on single posts only.',
		'before_widget' => '<div id="%1$s" class="widget %2$s">',
		'after_widget'  => '</div>',
		'before_title'  => '<h3 class="widget-title">',
		'after_title'   => '</h3>',
	] );
} );

add_action( 'wp_enqueue_scripts', function () {

	wp_enqueue_style(
		'snowbirdie-style',
		get_stylesheet_directory_uri() . '/style.css',
		[],
		wp_get_theme()->get( 'Version' )
	);
}, 20 );

add_action( 'snowbird_author_bio', function () {

	if ( is_single() ) {
		dynamic_sidebar( 'author-bio' );
	}
} );

add_action( 'snowbirdie.before_footer_widgets', function () {

	?>
	<div class="xf__container row clear">
		<div class="full first last">
			<?php the_widget( 'WP_Widget_Search' ); ?>
		</div>
	</div>
	<?php
} );
